File: jenjin-core/src/main/java/com/binarymonks/jj/time/TimeControls.java
package com.binarymonks.jj.time;

import com.badlogic.gdx.Gdx;
import com.binarymonks.jj.api.Time;
import com.binarymonks.jj.async.Function;

public class TimeControls implements Time {
    private static double fixedDelta = 1f / 60;
    private static float fixedDeltaFloat = 1f / 60;
    private static float realToGameRatio = 1.0f;
    private static double DELTA = 1f / 60;
    private static float DELTA_FLOAT = 1f / 60;
    private static double TIME = 0;
    private static TimeFunction timeFunction = TimeFunction.RATIO_TIME;
    private Scheduler scheduler = new Scheduler(this::getTime);


    @Override
    public double getDelta() {
        return DELTA;
    }

    @Override
    public float getDeltaFloat() {
        return DELTA_FLOAT;
    }

    public void update() {
        timeFunction.update(Gdx.graphics.getDeltaTime());
        TIME += DELTA;
        scheduler.update();
    }

    @Override
    public void setTimeFunction(TimeFunction function) {
        timeFunction = function;
    }

    @Override
    public double getTime() {
        return TIME;
    }

    @Override
    public double getFixedDelta() {
        return fixedDelta;
    }

    @Override
    public double getRatioFixedDelta() {
        return fixedDelta * realToGameRatio;
    }

    @Override
    public void setFixedDelta(double fixedTime) {
        TimeControls.fixedDelta = fixedTime;
        TimeControls.fixedDeltaFloat = new Double(fixedTime).floatValue();
    }

    @Override
    public double getRealToGameRatio() {
        return realToGameRatio;
    }

    @Override
    public void setRealToGameRatio(float realToGameRatio) {
        TimeControls.realToGameRatio = realToGameRatio;
    }

    @Override
    public TimeFunction getTimeFunction() {
        return timeFunction;
    }

    @Override
    public int scheduleInSeconds(Function function, float seconds, boolean repeat) {
        return scheduler.scheduleInSeconds(function, seconds, repeat);
    }

    @Override
    public void cancelScheduled(int scheduleID) {

    }

    public enum TimeFunction {
        REAL_TIME {
            @Override
            public void update(float realDelta) {
                DELTA = realDelta;
                DELTA_FLOAT = realDelta;
            }

        },
        FIXED_TIME {
            @Override
            public void update(float realDelta) {
                DELTA = fixedDelta;
                DELTA_FLOAT = fixedDeltaFloat;
            }

        },
        RATIO_TIME {
            @Override
            public void update(float realDelta) {
                DELTA = realToGameRatio * realDelta;
                DELTA_FLOAT = realToGameRatio * realDelta;
            }

        };

        public abstract void update(float realDelta);
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/time/TimeControls.kt
package com.binarymonks.jj.core.time

import com.badlogic.gdx.Gdx
import com.binarymonks.jj.core.api.TimeAPI

class TimeControls : TimeAPI {

    override val delta: Double
        get() = DELTA

    override val deltaFloat: Float
        get() = DELTA_FLOAT

    override val time: Double
        get() = TIME

    fun update() {
        timeFunction.update(Gdx.graphics.deltaTime)
        TIME += DELTA
    }

    fun setTimeFunction(function: TimeFunction) {
        timeFunction = function
    }

    fun getFixedDelta(): Double {
        return fixedDelta
    }

    val ratioFixedDelta: Double
        get() = fixedDelta * realToGameRatio

    fun setFixedDelta(fixedTime: Double) {
        TimeControls.fixedDelta = fixedTime
        TimeControls.fixedDeltaFloat = fixedTime.toFloat()
    }

    fun getRealToGameRatio(): Double {
        return realToGameRatio.toDouble()
    }

    fun setRealToGameRatio(realToGameRatio: Float) {
        TimeControls.realToGameRatio = realToGameRatio
    }

    fun getTimeFunction(): TimeFunction {
        return timeFunction
    }

    enum class TimeFunction {
        REAL_TIME {
            override fun update(realDelta: Float) {
                DELTA = realDelta.toDouble()
                DELTA_FLOAT = realDelta
            }

        },
        FIXED_TIME {
            override fun update(realDelta: Float) {
                DELTA = fixedDelta
                DELTA_FLOAT = fixedDeltaFloat
            }

        },
        RATIO_TIME {
            override fun update(realDelta: Float) {
                DELTA = (realToGameRatio * realDelta).toDouble()
                DELTA_FLOAT = realToGameRatio * realDelta
            }

        };

        abstract fun update(realDelta: Float)
    }

    companion object {
        private var fixedDelta = (1f / 60).toDouble()
        private var fixedDeltaFloat = 1f / 60
        private var realToGameRatio = 1.0f
        private var DELTA = (1f / 60).toDouble()
        private var DELTA_FLOAT = 1f / 60
        private var TIME = 0.0
        private var timeFunction = TimeFunction.RATIO_TIME
    }
}
